### PR TITLE
etcd: split TaskStatus into two structs

### PR DESCRIPTION
**Pull request by leoppro**
### What problem does this PR solve? 
reduce the probability of write conflicts about `TaskStatus`

### What is changed and how it works?
split `TaskStatus` into `TaskStatus` and `TaskPosition`
The processor write ResolvedTS and CheckpointTs into `TaskPosition`, and rarely write `TaskStatus`,
so we can reduce the probability of write conflicts about `TaskStatus`
### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: cdc/processor.go [hunk 1]
```diff
@@ -388,42 +390,43 @@ func (p *processor) localResolvedWorker(ctx context.Context) error {
...
-		p.handleTables(ctx, oldInfo, p.status, oldInfo.CheckPointTs)
+		p.handleTables(ctx, oldStatus, p.status, p.position.CheckPointTs)
 		syncTableNumGauge.WithLabelValues(p.changefeedID, p.captureID).Set(float64(len(p.status.TableInfos)))
 
-		if len(oldInfo.TableInfos) > len(p.status.TableInfos) {
-			// some table is removed, we will not both remove and add table in one operation.
-			// keep CheckpointTs and ResolvedTs as the old in memory cache one.
-			p.status.CheckPointTs = oldInfo.CheckPointTs
-			p.status.ResolvedTs = oldInfo.ResolvedTs
+		err = p.tsRWriter.WriteInfoIntoStorage(ctx)
```
Comment (suzaku) [inline cdc/processor.go]: Maybe we should move all the code that returns `Permanent` error outside the retried closure.

File: cdc/owner.go [hunk 2]
```diff
@@ -105,20 +106,21 @@ func (c *changeFeed) String() string {
 	return s
 }
 
-func (c *changeFeed) updateProcessorInfos(processInfos model.ProcessorsInfos) {
-	for cid, pinfo := range processInfos {
+func (c *changeFeed) updateProcessorInfos(processInfos model.ProcessorsInfos, position map[string]*model.TaskPosition) {
+	for cid, pinfo := range position {
```
Comment (amyangfei) [inline cdc/owner.go]: recommend to refine variable names in this function
s/position/positions
s/pinfo/position
s/oldPinfo/oldPosition

File: cdc/owner.go [hunk 3]
```diff
@@ -438,6 +440,16 @@ func (o *ownerImpl) handleMarkdownProcessor(ctx context.Context) {
 			remainProcs = append(remainProcs, snap)
 			continue
 		}
+		err = kv.DeleteTaskPosition(ctx, o.etcdClient, snap.CfID, snap.CaptureID)
+		if err != nil {
+			log.Warn("failed to delete processor info",
```
Comment (amyangfei) [inline cdc/owner.go]:
```suggestion
			log.Warn("failed to delete task position",
```
similar suggestion in L435, s/processor info/task status

File: cdc/owner.go [hunk 4]
```diff
@@ -61,7 +61,7 @@ type OwnerDDLHandler interface {
 // ChangeFeedRWriter defines the Reader and Writer for changeFeed
 type ChangeFeedRWriter interface {
 	// Read the changefeed info from storage such as etcd.
-	Read(ctx context.Context) (map[model.ChangeFeedID]*model.ChangeFeedInfo, map[model.ChangeFeedID]*model.ChangeFeedStatus, map[model.ChangeFeedID]model.ProcessorsInfos, error)
+	Read(ctx context.Context) (map[model.ChangeFeedID]*model.ChangeFeedInfo, map[model.ChangeFeedID]*model.ChangeFeedStatus, map[model.ChangeFeedID]model.ProcessorsInfos, map[model.ChangeFeedID]map[model.CaptureID]*model.TaskPosition, error)
```
[the remark this answers is not in the document]
Comment (leoppro) [reply cdc/owner.go]: yes, I solved this problem in #255

File: cdc/owner.go [hunk 5]
```diff
@@ -523,9 +543,10 @@ func (o *ownerImpl) newChangeFeed(id model.ChangeFeedID, processorsInfos model.P
 	ddlHandler := newDDLHandler(o.pdClient, checkpointTs)
 
 	existingTables := make(map[uint64]uint64)
-	for _, taskStatus := range processorsInfos {
+	for captureID, taskStatus := range processorsInfos {
+		pos := taskPositions[captureID]
```
Comment (amyangfei) [inline cdc/owner.go]: should we check whether `pos` is nil

File: cdc/processor.go (hunk 1 above)
Comment (leoppro) [reply cdc/processor.go]: why? I think we should retry when the write operation is conflict

Comment (leoppro) [reply cdc/processor.go]: I see

File: cdc/kv/etcd.go [hunk 6]
```diff
@@ -169,10 +179,34 @@ func (c CDCEtcdClient) SaveChangeFeedInfo(ctx context.Context, info *model.Chang
...
 
+// GetAllTaskPositions queries all task positions of a changefeed, and returns a map
+// mapping from captureID to TaskPositions
+func (c CDCEtcdClient) GetAllTaskPositions(ctx context.Context, changefeedID string, opts ...clientv3.OpOption) (map[string]*model.TaskPosition, error) {
+	key := GetEtcdKeyTaskPositionList(changefeedID)
+	resp, err := c.Client.Get(ctx, key, append([]clientv3.OpOption{clientv3.WithPrefix()}, opts...)...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	positions := make(map[string]*model.TaskPosition, resp.Count)
```
Comment (suzaku) [inline cdc/kv/etcd.go]: It seems `model.TaskPosition` is a tiny struct, we can use value instead of pointer for better performance?

Comment (leoppro) [reply cdc/kv/etcd.go]: no, I think we should use pointer, If we try to update `ResolvedTS` or `CheckpointTS`, we should make sure that the new value will be set to `positionsMap`

**Pull request merged**